fms/modules/linear.py

```diff
@@ -0,0 +1,169 @@
+import torch
+import torch.nn as nn
+from typing import Callable, Mapping, Any
+
+
+__type_factory_map: Mapping[str, Callable] = {}
+__type_sharding_map : Mapping[str, Callable] = {}
+
+
+def register_linear_type_to_module_map(linear_type: str, factory: Callable) -> None:
+    if linear_type in __type_factory_map:
+        raise KeyError(
+            f"Module mapping of linear type `{linear_type}` already registered"
+        )
+    __type_factory_map[linear_type] = factory
+
+
+def register_linear_type_to_sharding_map(linear_type: str, factory: Callable) -> None:
+    if linear_type in __type_sharding_map:
+        raise KeyError(
+            f"Sharding map of linear type `{linear_type}` already registered"
+        )
+    __type_sharding_map[linear_type] = factory
+
+
+def get_all_linear_type_to_sharding_maps() -> dict[str, Callable]:
+    return __type_sharding_map
+
+
+def get_linear_type(linear_config: Mapping[str, Any] | None) -> str:
+    if not linear_config:
+        return "torch_linear"
+
+    linear_type = linear_config.get("linear_type", None)
+    if not linear_type:
+        return "torch_linear"
+    if not isinstance(linear_type, str):
+        raise TypeError("linear_type in linear_config must be string")
+    if linear_type.lower() not in __type_factory_map:
+        raise ValueError(f"Unsupported linear_type `{linear_type}` in linear_config")
+
+    return linear_type.lower()
+
+
+def get_linear(
+    in_features: int,
+    out_features: int,
+    bias: bool,
+    linear_config: Mapping[str, Any] | None = None,
+) -> nn.Module:
+    linear_type = get_linear_type(linear_config)
+
+    # TODO: how to merge these calls that get different arguments?
+    if linear_type in __type_factory_map:
+        if linear_type == "torch_linear":
+            return __type_factory_map[linear_type](in_features, out_features, bias)
+        else:
+            return __type_factory_map[linear_type](
+                in_features, out_features, bias, linear_config
+            )
+    else:
+        raise KeyError(f"Unsupported linear type `{linear_type}`")
+
+
+def shard_base_linear(
```

Comment:
This function, while called "base", it contains specialized handling of gptq. We need to separate specialization from generalization so that the general parts can be reused in fp8

Reply:
this has been resolved